pkg/controller/core/workload_controller.go

```diff
@@ -137,32 +137,21 @@ func (r *WorkloadReconciler) Reconcile(ctx context.Context, req ctrl.Request) (c
 		return ctrl.Result{}, nil
 	}
 
-	if rejectedChecks := workload.GetRejectedChecks(&wl); len(rejectedChecks) > 0 {
-		// Finish the workload
-		log.V(3).Info("Workload has Rejected admission checks, Finish with failure")
-		err := workload.UpdateStatus(ctx, r.client, &wl, kueue.WorkloadFinished,
-			metav1.ConditionTrue,
-			"AdmissionChecksRejected",
-			fmt.Sprintf("Admission checks %v are rejected", rejectedChecks),
-			constants.KueueName)
-		return ctrl.Result{}, err
-	}
-
 	cqName, cqOk := r.queues.ClusterQueueForWorkload(&wl)
 	if cqOk {
 		if updated, err := r.reconcileSyncAdmissionChecks(ctx, &wl, cqName); updated || err != nil {
 			return ctrl.Result{}, err
 		}
 	}
 
-	if workload.SyncAdmittedCondition(&wl) {
+	// If the workload is not admitted update its admitted condition if necessary.
+	if !workload.IsAdmitted(&wl) && workload.SyncAdmittedCondition(&wl) {
```

Comment:
Is this change needed, or just a performance optimization? If needed, does the order matter?

Also, is it correct, I'm asking because if the workload was already admitted, then (IIUC) we don't enter the branch, we also skip `if workload.HasQuotaReservation(&wl) {`, then we enter `if rejectedChecks := workload.GetRejectedChecks(&wl); len(rejectedChecks) > 0 {`. So we would mark the workload as finished.
I think the idea was to go via eviction rather than marking the workload as finished? Can you explain what is the code path leading to workload eviction in that scenario now?

Reply:
It is needed, and the order is important since  `SyncAdmittedCondition` can update the `Admitted` cind. in the wl.

If the workload is admitted, it will skip this  and just check if the eviction is needed.

Reply:
I checked that the integration test passes when the check `!workload.IsAdmitted(&wl)` is removed. Can we add extend the integration test to demonstrate its importance?

Reply:
> (If we don't keep this the Admitted condition is cleared before the eviction starts)

Why does that matter? 

To be clear I'm fine with having Admitted=True and Evicted=True, I'm also fine with clearing the Admitted before adding Evicted=True. Trying to understand what is the difference, to avoid unnecessary code complications.

Reply:
comment updated. 

> To be clear I'm fine with having Admitted=True and Evicted=True, I'm also fine with clearing the Admitted before adding Evicted=True. Trying to understand what is the difference, to avoid unnecessary code complications.

It's more logical in my opinion and is the same flow as in case of premonition based eviction.

Reply:
sgtm, I buy the argument of consistent flow with preemption-based eviction